README.md
# ИДЗ по АВС №2
### Цейтин Андрей Владимирович БПИ217
# <br> Отчет
## <br> На 4 балла:
### 1) Приведено решение задачи на C (с использованием двух единиц компиляции):
* [main.c](https://github.com/CehhGhost/ABC2/blob/main/C%20code/main.c)
* [get_maxsymb_and_minsymb.c](https://github.com/CehhGhost/ABC2/blob/main/C%20code/get_maxsymb_and_minsymb.c)
### 2)  В полученную ассемблерную программу, откомпилированную без оптимизирующих и отладочных опций, добавлены комментарии, поясняющие
эквивалентное представление переменных в программе на C:
* [main.s](https://github.com/CehhGhost/ABC2/blob/main/Assembled/main.s)
* [get_maxsymb_and_minsymb.s](https://github.com/CehhGhost/ABC2/blob/main/Assembled/get_maxsymb_and_minsymb.s)
### 3) Из ассемблерной программы убраны лишние макросы за счет использования соответствующих аргументов командной строки и/или за счет
ручного редактирования исходного текста ассемблерной программы:
* [main_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/main_updated.s)
* [get_maxsymb_and_minsymb_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/get_maxsymb_and_minsymb_updated.s)
#### [Список изменений](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/Changes.md)
### 4) тестирование
> [ссылка на тесты](https://github.com/CehhGhost/ABC2/blob/main/tests.md)
> <br> При просмотре тестов можно заметить, что программа до модификаций и после работает одинаково и без ошибок => модификации установленны верно
## <br> На 5 баллов:
### 1) В реализованной программе использовать функции с передачей данных через параметры и использованны локальные переменные:
В программе используются локальные переменные (например int size), присутствует одна функция с передачей данных через параметры:
* get_maxsymb_and_minsymb(const char *line, int size, char* min, char* max)
### 2) В программу добавленны комментарии, описывающие передачу фактических параметров, перенос возвращаемого результата и связь между параметрами языка Си и регистрами (стеком)
* [main_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/main_updated.s)
* [get_maxsymb_and_minsymb_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/get_maxsymb_and_minsymb_updated.s)
## <br> На 6 баллов:
### 1) Рефакторинг программы на ассемблере за счет максимального использования регистров процессора:
#### В main:
> * r12d
> * r14d
#### В get_maxsymb_and_minsymb:
> * r14d
### 2) В программу добавленны комментарии, поясняющие эквивалентное использование регистров вместо переменных исходной программы на C
* [main_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/main_updated.s)
* [get_maxsymb_and_minsymb_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/get_maxsymb_and_minsymb_updated.s)
### 3) Тестирование
> [ссылка на тесты](https://github.com/CehhGhost/ABC2/blob/main/tests.md)
> <br> При просмотре тестов можно заметить, что программа до модификаций и после работает одинаково и без ошибок => модификации установленны верно
## <br> На 7 баллов:
### 1) Реализация программы на ассемблере, полученной после рефакторинга, в виде двух или более единиц компиляции (main_updated.s, get_maxsymb_and_minsymb_updated.s):
* [main_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/main_updated.s)
* [get_maxsymb_and_minsymb_updated.s](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/get_maxsymb_and_minsymb_updated.s)
### 2) Файл с входными данными - [input.txt](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/input.txt), файл с выходными данными - [output.txt](https://github.com/CehhGhost/ABC2/blob/main/Assembeld%20updated/output.txt)
### 3) Тестирование
> [ссылка на тесты](https://github.com/CehhGhost/ABC2/blob/main/tests.md)
> <br> При просмотре тестов можно заметить, что программа корректно считывает из входного файла и записывает в выходной файл


Assembeld updated/Changes.md
# Модификации
## Был использован аргумент командной строки: 
~~~
gcc -masm=intel \
-fno-asynchronous-unwind-tables \
-fno-jump-tables \
-fno-stack-protector \
-fno-exceptions \
./make_array.c \
-S -o ./make_array_edited.s
~~~
## <br> Также удалены бесполезный присваивания из main.s:
> \- mov	rax, rsp
> <br> - mov	rbx, rax
> <br> + mov	rbx, rsp

> \- mov	rax, DWORD PTR -88[rbp]
> <br> - mov	rdi, rax
> <br> + mov	rdi, DWORD PTR -88[rbp]

> \- mov	rax, DWORD PTR -104[rbp]
> <br> - mov	rdi, rax
> <br> + mov	rdi, DWORD PTR -104[rbp]

> \- mov	eax, DWORD PTR -56[rbp]
> <br> - mov	esi, eax
> <br> + mov	esi, DWORD PTR -56[rbp]

> \- mov	rax, DWORD PTR -72[rbp]
> <br> - mov	rdi, rax
> <br> + mov	rdi, DWORD PTR -72[rbp]

> \- mov	eax, DWORD PTR -52[rbp]
> <br> - mov	esi, eax
> <br> + mov	esi, DWORD PTR -52[rbp]

## <br> Так же во всех файлах удалены:
> endbr64


C code/main.c
#include <stdio.h>
#include <stdlib.h>

extern void get_maxsymb_and_minsymb(const char *line, int size, char* min, char* max);

int main(int argc, char *argv[]) {
    int size;
    int flag;
    int seed;
    char min;
    char max;
    if (argc == 2) {
        size = atoi(argv[1]);
        char line[size];
        FILE *input = fopen("input.txt", "r");
        fgets(line, size + 1, input);
        fclose(input);
        FILE *output = fopen("output.txt", "w");
        get_maxsymb_and_minsymb(line, size, &min, &max);
        fprintf(output, "Minimum symb: %c\r\n", min);
        fprintf(output, "Maximum symb: %c\r\n", max);
        fclose(output);
        return 0;
    }
    printf("Set size of your line(>0 && <= %d)\r\n", 100);
    scanf("%d", &size);
    if (size < 1 || size > 100) {
        printf("Size is out of range\r\n");
        return 1;
    }
    char line[size];
    printf("Choose input: manual or random(0 = manual; other = random):\r\n");
    scanf("%d", &flag);
    if (flag == 0) {
        printf("\n");
        for (int i = 0; i < size; ++i) {
            printf("Line[%d] =", i);
            scanf("%s", line + i);
        }
        get_maxsymb_and_minsymb(line, size, &min, &max);
        printf("Minimum symb: %c\r\n", min);
        printf("Maximum symb: %c\r\n", max);
        return 0;
    }
    printf("Choose seed generation(>0 && <= 100):\r\n");
    scanf("%d", &seed);
    if (seed < 1 || seed > 100) {
        printf("Seed is out of range\r\n");
        return 1;
    }
    srand(seed);
    printf("Result Line: \r\n");
    for (int i = 0; i < size; ++i) {
        line[i] = (char)((int)'A' + rand() % ('Z' - 'A'));
        printf("Line[%d] = %c\r\n", i, *(line + i));
    }
    printf("\r\n");
    get_maxsymb_and_minsymb(line, size, &min, &max);
    printf("Minimum symb: %c\r\n", min);
    printf("Maximum symb: %c\r\n", max);
    return 0;
}


Assembled/get_maxsymb_and_minsymb.s
	.file	"get_maxsymb_and_minsymb.c"
	.intel_syntax noprefix
	.text
	.globl	get_maxsymb_and_minsymb
	.type	get_maxsymb_and_minsymb, @function
get_maxsymb_and_minsymb:
.LFB0:
	.cfi_startproc
	endbr64
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	mov	rbp, rsp
	.cfi_def_cfa_register 6
	mov	QWORD PTR -24[rbp], rdi		# line (= -24)
	mov	DWORD PTR -28[rbp], esi		# size (= -28)
	mov	QWORD PTR -40[rbp], rdx		# *min (= -40)
	mov	QWORD PTR -48[rbp], rcx		# *max (= -48)
	mov	rax, QWORD PTR -24[rbp]
	movzx	edx, BYTE PTR [rax]
	mov	rax, QWORD PTR -40[rbp]
	mov	BYTE PTR [rax], dl
	mov	rax, QWORD PTR -24[rbp]
	movzx	edx, BYTE PTR [rax]
	mov	rax, QWORD PTR -48[rbp]
	mov	BYTE PTR [rax], dl
	mov	DWORD PTR -4[rbp], 1
	jmp	.L2
.L5:
	mov	eax, DWORD PTR -4[rbp]
	movsx	rdx, eax
	mov	rax, QWORD PTR -24[rbp]
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	mov	BYTE PTR -5[rbp], al
	mov	rax, QWORD PTR -48[rbp]
	movzx	eax, BYTE PTR [rax]
	cmp	BYTE PTR -5[rbp], al
	jle	.L3
	mov	rax, QWORD PTR -48[rbp]
	movzx	edx, BYTE PTR -5[rbp]
	mov	BYTE PTR [rax], dl
.L3:
	mov	rax, QWORD PTR -40[rbp]
	movzx	eax, BYTE PTR [rax]
	cmp	BYTE PTR -5[rbp], al
	jge	.L4
	mov	rax, QWORD PTR -40[rbp]
	movzx	edx, BYTE PTR -5[rbp]
	mov	BYTE PTR [rax], dl
.L4:
	add	DWORD PTR -4[rbp], 1
.L2:
	mov	eax, DWORD PTR -4[rbp]
	cmp	eax, DWORD PTR -28[rbp]
	jl	.L5
	nop
	nop
	pop	rbp
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE0:
	.size	get_maxsymb_and_minsymb, .-get_maxsymb_and_minsymb
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4:


Assembeld updated/get_maxsymb_and_minsymb_updated.s
	.file	"get_maxsymb_and_minsymb.c"
	.intel_syntax noprefix
	.text
	.globl	get_maxsymb_and_minsymb
	.type	get_maxsymb_and_minsymb, @function
get_maxsymb_and_minsymb:
	push	rbp
	mov	rbp, rsp
	mov	QWORD PTR -24[rbp], rdi		# line (= -24)
	mov	DWORD PTR -28[rbp], esi		# size (= -28)
	mov	QWORD PTR -40[rbp], rdx		# *min (= -40)
	mov	QWORD PTR -48[rbp], rcx		# *max (= -48)
	mov	rax, QWORD PTR -24[rbp]
	movzx	edx, BYTE PTR [rax]
	mov	rax, QWORD PTR -40[rbp]
	mov	BYTE PTR [rax], dl
	mov	rax, QWORD PTR -24[rbp]
	movzx	edx, BYTE PTR [rax]
	mov	rax, QWORD PTR -48[rbp]
	mov	BYTE PTR [rax], dl
	mov	r14d, 1				# r14d = i = 1 в цикле for()
	jmp	.L2
.L5:	# цикл for()
	mov	eax, r14d			# eax = i
	movsx	rdx, eax
	mov	rax, QWORD PTR -24[rbp]
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	mov	BYTE PTR -5[rbp], al
	mov	rax, QWORD PTR -48[rbp]
	movzx	eax, BYTE PTR [rax]
	cmp	BYTE PTR -5[rbp], al		# if (symb > *max)
	jle	.L3
	mov	rax, QWORD PTR -48[rbp]
	movzx	edx, BYTE PTR -5[rbp]
	mov	BYTE PTR [rax], dl		# *max = symb
.L3:
	mov	rax, QWORD PTR -40[rbp]
	movzx	eax, BYTE PTR [rax]
	cmp	BYTE PTR -5[rbp], al		# if (symb < *min)
	jge	.L4
	mov	rax, QWORD PTR -40[rbp]
	movzx	edx, BYTE PTR -5[rbp]
	mov	BYTE PTR [rax], dl		# *min = symb
.L4:
	add	r14d, 1				# ++i
.L2:
	mov	eax, r14d			# eax = i
	cmp	eax, DWORD PTR -28[rbp]		# i < size
	jl	.L5
	nop
	nop
	pop	rbp
	ret
	.size	get_maxsymb_and_minsymb, .-get_maxsymb_and_minsymb
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4:


Assembled/main.s
	.file	"main.c"
	.intel_syntax noprefix
	.text
	.section	.rodata
.LC0:
	.string	"r"
.LC1:
	.string	"input.txt"
.LC2:
	.string	"w"
.LC3:
	.string	"output.txt"
.LC4:
	.string	"Minimum symb: %c\r\n"
.LC5:
	.string	"Maximum symb: %c\r\n"
	.align 8
.LC6:
	.string	"Set size of your line(>0 && <= %d)\r\n"
.LC7:
	.string	"%d"
.LC8:
	.string	"Size is out of range\r"
	.align 8
.LC9:
	.string	"Choose input: manual or random(0 = manual; other = random):\r"
.LC10:
	.string	"Line[%d] ="
.LC11:
	.string	"%s"
	.align 8
.LC12:
	.string	"Choose seed generation(>0 && <= 100):\r"
.LC13:
	.string	"Seed is out of range\r"
.LC14:
	.string	"Result Line: \r"
.LC15:
	.string	"Line[%d] = %c\r\n"
.LC16:
	.string	"\r"
	.text
	.globl	main
	.type	main, @function
main:
.LFB6:
	.cfi_startproc
	endbr64
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	mov	rbp, rsp
	.cfi_def_cfa_register 6
	push	r15
	push	r14
	push	r13
	push	r12
	push	rbx
	sub	rsp, 136
	.cfi_offset 15, -24
	.cfi_offset 14, -32
	.cfi_offset 13, -40
	.cfi_offset 12, -48
	.cfi_offset 3, -56
	mov	DWORD PTR -132[rbp], edi	# int argc (= -132)
	mov	QWORD PTR -144[rbp], rsi	# char *argv[] (= -144)
	mov	rax, QWORD PTR fs:40
	mov	QWORD PTR -56[rbp], rax
	xor	eax, eax
	mov	rax, rsp
	mov	rbx, rax
	cmp	DWORD PTR -132[rbp], 2		# сравнить argc с 2 (argc == 2)
	jne	.L2				# if (argc == 2) {...} else
	mov	rax, rsp
	mov	r12, rax
	mov	rax, QWORD PTR -144[rbp]	# сохраняем argv[0]
	add	rax, 8				# для чтения с командной строки сдвигаем на argv[0]
	mov	rax, QWORD PTR [rax]
	mov	rdi, rax
	call	atoi@PLT
	mov	DWORD PTR -124[rbp], eax	# int size = eax (= -124)
	mov	eax, DWORD PTR -124[rbp]	# eax = size
	movsx	rdx, eax
	sub	rdx, 1
	mov	QWORD PTR -88[rbp], rdx
	movsx	rdx, eax
	mov	QWORD PTR -160[rbp], rdx
	mov	QWORD PTR -152[rbp], 0
	movsx	rdx, eax
	mov	QWORD PTR -176[rbp], rdx
	mov	QWORD PTR -168[rbp], 0
	cdqe
	mov	edx, 16
	sub	rdx, 1
	add	rax, rdx
	mov	esi, 16
	mov	edx, 0
	div	rsi
	imul	rax, rax, 16
	mov	rdx, rax
	and	rdx, -4096
	mov	rcx, rsp
	sub	rcx, rdx
	mov	rdx, rcx
.L3:
	cmp	rsp, rdx
	je	.L4
	sub	rsp, 4096
	or	QWORD PTR 4088[rsp], 0
	jmp	.L3
.L4:
	mov	rdx, rax
	and	edx, 4095
	sub	rsp, rdx
	mov	rdx, rax
	and	edx, 4095
	test	rdx, rdx
	je	.L5
	and	eax, 4095
	sub	rax, 8
	add	rax, rsp
	or	QWORD PTR [rax], 0
.L5:
	mov	rax, rsp
	add	rax, 0
	mov	QWORD PTR -80[rbp], rax		# int line[] (= -80)
	lea	rsi, .LC0[rip]			# ссылка на 'r'
	lea	rdi, .LC1[rip]			# ссылка на 'input.txt'
	call	fopen@PLT
	mov	QWORD PTR -72[rbp], rax		# FILE *input (= -72)
	mov	eax, DWORD PTR -124[rbp]	# size для передачи в fgets(line, size + 1, input)
	lea	ecx, 1[rax]			# size + 1
	mov	rdx, QWORD PTR -72[rbp]		# input для fgets(line, size + 1, input)
	mov	rax, QWORD PTR -80[rbp]		# line для передачи в fgets(line, size + 1, input)
	mov	esi, ecx
	mov	rdi, rax
	call	fgets@PLT
	mov	rax, QWORD PTR -72[rbp]
	mov	rdi, rax
	call	fclose@PLT
	lea	rsi, .LC2[rip]			# ссылка на 'w'
	lea	rdi, .LC3[rip]			# cсылка на 'output.txt'
	call	fopen@PLT
	mov	QWORD PTR -64[rbp], rax		# FILE *output (= -64)
	mov	esi, DWORD PTR -124[rbp]	# size для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	lea	rcx, -125[rbp]			# min для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	lea	rdx, -126[rbp]			# max для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	mov	rax, QWORD PTR -80[rbp]		# line для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	mov	rdi, rax
	call	get_maxsymb_and_minsymb@PLT
	movzx	eax, BYTE PTR -126[rbp]		# min для передачи в fprintf(output, "Minimum symb: %c\r\n", min)
	movsx	edx, al
	mov	rax, QWORD PTR -64[rbp]		# output для передачи в fprintf(output, "Minimum symb: %c\r\n", min)
	lea	rsi, .LC4[rip]
	mov	rdi, rax
	mov	eax, 0
	call	fprintf@PLT
	movzx	eax, BYTE PTR -125[rbp]		# max для передачи в fprintf(output, "Maximum symb: %c\r\n", max)
	movsx	edx, al
	mov	rax, QWORD PTR -64[rbp]		# output для передачи в fprintf(output, "Maximum symb: %c\r\n", max)
	lea	rsi, .LC5[rip]
	mov	rdi, rax
	mov	eax, 0
	call	fprintf@PLT
	mov	rax, QWORD PTR -64[rbp]
	mov	rdi, rax
	call	fclose@PLT
	mov	eax, 0
	mov	rsp, r12
	jmp	.L6
.L2:	# if (argc != 2)
	mov	esi, 100
	lea	rdi, .LC6[rip]
	mov	eax, 0
	call	printf@PLT			# printf("Set size of your line(>0 && <= %d)\r\n", 100)
	lea	rax, -124[rbp]			# size для передачи в scanf("%d", &size);
	mov	rsi, rax
	lea	rdi, .LC7[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT
	mov	eax, DWORD PTR -124[rbp]
	test	eax, eax
	jle	.L7
	mov	eax, DWORD PTR -124[rbp]
	cmp	eax, 100
	jle	.L8
.L7:
	lea	rdi, .LC8[rip]
	call	puts@PLT
	mov	eax, 1
	jmp	.L6
.L8:
	mov	eax, DWORD PTR -124[rbp]
	movsx	rdx, eax
	sub	rdx, 1
	mov	QWORD PTR -104[rbp], rdx
	movsx	rdx, eax
	mov	r14, rdx
	mov	r15d, 0
	movsx	rdx, eax
	mov	r12, rdx
	mov	r13d, 0
	cdqe
	mov	edx, 16
	sub	rdx, 1
	add	rax, rdx
	mov	esi, 16
	mov	edx, 0
	div	rsi
	imul	rax, rax, 16
	mov	rdx, rax
	and	rdx, -4096
	mov	rcx, rsp
	sub	rcx, rdx
	mov	rdx, rcx
.L9:
	cmp	rsp, rdx
	je	.L10
	sub	rsp, 4096
	or	QWORD PTR 4088[rsp], 0
	jmp	.L9
.L10:
	mov	rdx, rax
	and	edx, 4095
	sub	rsp, rdx
	mov	rdx, rax
	and	edx, 4095
	test	rdx, rdx
	je	.L11
	and	eax, 4095
	sub	rax, 8
	add	rax, rsp
	or	QWORD PTR [rax], 0
.L11:
	mov	rax, rsp
	add	rax, 0
	mov	QWORD PTR -96[rbp], rax
	lea	rdi, .LC9[rip]
	call	puts@PLT
	lea	rax, -120[rbp]
	mov	rsi, rax
	lea	rdi, .LC7[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT
	mov	eax, DWORD PTR -120[rbp]
	test	eax, eax
	jne	.L12
	mov	edi, 10
	call	putchar@PLT
	mov	DWORD PTR -108[rbp], 0
	jmp	.L13
.L14:
	mov	eax, DWORD PTR -108[rbp]
	mov	esi, eax
	lea	rdi, .LC10[rip]
	mov	eax, 0
	call	printf@PLT
	mov	eax, DWORD PTR -108[rbp]
	movsx	rdx, eax
	mov	rax, QWORD PTR -96[rbp]
	add	rax, rdx
	mov	rsi, rax
	lea	rdi, .LC11[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT
	add	DWORD PTR -108[rbp], 1
.L13:
	mov	eax, DWORD PTR -124[rbp]
	cmp	DWORD PTR -108[rbp], eax
	jl	.L14
	mov	esi, DWORD PTR -124[rbp]	# size для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	lea	rcx, -125[rbp]			# min для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	lea	rdx, -126[rbp]			# max для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	mov	rax, QWORD PTR -96[rbp]		# line для передачи в get_maxsymb_and_minsymb(line, size, &min, &max)
	mov	rdi, rax
	call	get_maxsymb_and_minsymb@PLT
	movzx	eax, BYTE PTR -126[rbp]		# min для передачи в printf("Minimum symb: %c\r\n", min)
	movsx	eax, al
	mov	esi, eax
	lea	rdi, .LC4[rip]
	mov	eax, 0
	call	printf@PLT
	movzx	eax, BYTE PTR -125[rbp]		# max для передачи в printf("Maximum symb: %c\r\n", max);
	movsx	eax, al
	mov	esi, eax
	lea	rdi, .LC5[rip]
	mov	eax, 0
	call	printf@PLT
	mov	eax, 0
	jmp	.L6
.L12:
	lea	rdi, .LC12[rip]
	call	puts@PLT
	lea	rax, -116[rbp]
	mov	rsi, rax
	lea	rdi, .LC7[rip]
	mov	eax, 0
	call	__isoc99_scanf@PLT
	mov	eax, DWORD PTR -116[rbp]
	test	eax, eax
	jle	.L15
	mov	eax, DWORD PTR -116[rbp]
	cmp	eax, 100
	jle	.L16
.L15:
	lea	rdi, .LC13[rip]
	call	puts@PLT
	mov	eax, 1
	jmp	.L6
.L16:
	mov	eax, DWORD PTR -116[rbp]
	mov	edi, eax
	call	srand@PLT
	lea	rdi, .LC14[rip]
	call	puts@PLT
	mov	DWORD PTR -112[rbp], 0
	jmp	.L17
.L18:
	call	rand@PLT
	mov	edx, eax
	movsx	rax, edx
	imul	rax, rax, 1374389535
	shr	rax, 32
	mov	ecx, eax
	sar	ecx, 3
	mov	eax, edx
	sar	eax, 31
	sub	ecx, eax
	mov	eax, ecx
	sal	eax, 2
	add	eax, ecx
	lea	ecx, 0[0+rax*4]
	add	eax, ecx
	mov	ecx, edx
	sub	ecx, eax
	mov	eax, ecx
	add	eax, 65
	mov	ecx, eax
	mov	rdx, QWORD PTR -96[rbp]
	mov	eax, DWORD PTR -112[rbp]
	cdqe
	mov	BYTE PTR [rdx+rax], cl
	mov	eax, DWORD PTR -112[rbp]
	movsx	rdx, eax
	mov	rax, QWORD PTR -96[rbp]
	add	rax, rdx
	movzx	eax, BYTE PTR [rax]
	movsx	edx, al
	mov	eax, DWORD PTR -112[rbp]
	mov	esi, eax
	lea	rdi, .LC15[rip]
	mov	eax, 0
	call	printf@PLT
	add	DWORD PTR -112[rbp], 1
.L17:
	mov	eax, DWORD PTR -124[rbp]
	cmp	DWORD PTR -112[rbp], eax
	jl	.L18
	lea	rdi, .LC16[rip]
	call	puts@PLT
	mov	esi, DWORD PTR -124[rbp]
	lea	rcx, -125[rbp]
	lea	rdx, -126[rbp]
	mov	rax, QWORD PTR -96[rbp]
	mov	rdi, rax
	call	get_maxsymb_and_minsymb@PLT
	movzx	eax, BYTE PTR -126[rbp]
	movsx	eax, al
	mov	esi, eax
	lea	rdi, .LC4[rip]
	mov	eax, 0
	call	printf@PLT
	movzx	eax, BYTE PTR -125[rbp]
	movsx	eax, al
	mov	esi, eax
	lea	rdi, .LC5[rip]
	mov	eax, 0
	call	printf@PLT
	mov	eax, 0
.L6:
	mov	rsp, rbx
	mov	rbx, QWORD PTR -56[rbp]
	xor	rbx, QWORD PTR fs:40
	je	.L20
	call	__stack_chk_fail@PLT
.L20:
	lea	rsp, -40[rbp]
	pop	rbx
	pop	r12
	pop	r13
	pop	r14
	pop	r15
	pop	rbp
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE6:
	.size	main, .-main
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4:
